cmake/ament_doxygen_generate.cmake

```diff
@@ -15,102 +15,146 @@
+    REALPATH BASE_DIR "${CMAKE_CURRENT_SOURCE_DIR}"
+  )
 
-  foreach(dep ${args_DEPENDENCIES})
-    ament_index_has_resource(has_tag doxygen_tags "${dep}")
-    if(has_tag)
-      ament_index_get_resource(tag doxygen_tags "${dep}")
-      list(GET tag 0 tagfile)
-      list(GET tag 1 htmldir)
-      file(RELATIVE_PATH htmldir "${OUTPUT_DIRECTORY}/html" "${htmldir}")
-      set(TAGFILES "${TAGFILES} \"${tagfile}=${htmldir}\"")
-    else()
-      message(WARNING "A Doxygen tagfile for ${dep} could not be found, ignoring")
-    endif()
-  endforeach()
-  set(GENERATE_TAGFILE "${OUTPUT_DIRECTORY}/${PROJECT_NAME}.tag")
-
-  configure_file("${CONFIG_DIRECTORY}/Doxyfile.in" "${BUILD_DIRECTORY}/Doxyfile" @ONLY)
+  if (args_STANDALONE)
+    ament_doxygen_aggregate_external(${target_name}_dependencies
+      EXTERNAL_PROJECTS ${args_DEPENDENCIES}
+      CONFIG_DIRECTORY "${args_CONFIG_DIRECTORY}/deps"
+      BUILD_DIRECTORY "${args_BUILD_DIRECTORY}/html/deps"
+    )
 
-  if (args_CONFIG_OVERLAY)
-    string_ends_with("${args_CONFIG_OVERLAY}" ".in" is_template)
-    if(is_template)
-      get_filename_component(overlay_filename "${args_CONFIG_OVERLAY}" NAME)
-      # cut off .in extension
-      string(LENGTH "${overlay_filename}" length)
-      math(EXPR offset "${length} - 3")
-      string(SUBSTRING "${overlay_filename}" 0 ${offset} overlay_filename)
-      configure_file(
-        "${args_CONFIG_OVERLAY}"
-        "${BUILD_DIRECTORY}/${overlay_filename}"
-        @ONLY
-      )
-      set(args_CONFIG_OVERLAY "${BUILD_DIRECTORY}/${overlay_filename}")
-    endif()
-    file(READ "${args_CONFIG_OVERLAY}" overlay_content)
-    file(APPEND "${BUILD_DIRECTORY}/Doxyfile" "${overlay_content}")
+    ament_doxygen_resolve_tags(tags
+      TAGS ${${target_name}_dependencies_TAGS}
+      BASE_DIRECTORY "${args_BUILD_DIRECTORY}/html"
+    )
+  else()
+    ament_doxygen_find_external(tags
+      EXTERNAL_PROJECTS ${args_DEPENDENCIES}
+      BASE_DIRECTORY "${args_BUILD_DIRECTORY}/html"
+    )
   endif()
 
-  add_custom_target(doxygen_${PROJECT_NAME} ALL
-    COMMAND ${DOXYGEN_EXECUTABLE} "${BUILD_DIRECTORY}/Doxyfile"
-    VERBATIM
+  set(doxyfile "${args_CONFIG_DIRECTORY}/Doxyfile")
+  ament_doxygen_generate_configuration("${doxyfile}"
+    CONFIG_OVERLAY "${args_CONFIG_OVERLAY}"
+    CONFIG_DIRECTORY "${args_CONFIG_DIRECTORY}"
+    BUILD_DIRECTORY "${args_BUILD_DIRECTORY}"
+    INPUT_DIRECTORY "${args_INPUT_DIRECTORY}"
+    PROJECT_NAME "${args_PROJECT_NAME}"
+    OUTPUT_TAGFILE "${args_PROJECT_NAME}.tag"
+    INPUT_TAGS ${tags}
   )
-  ament_index_register_resource(doxygen_tags
-    PACKAGE_NAME "${PROJECT_NAME}"
-    CONTENT "${GENERATE_TAGFILE};${OUTPUT_DIRECTORY}/html"
+
+  add_custom_target(${target_name} ALL
+    COMMAND ${DOXYGEN_EXECUTABLE} "${doxyfile}"
+    VERBATIM
   )
+
+  if (TARGET ${target_name}_dependencies)
+    add_dependencies(${target_name} ${target_name}_dependencies)
+  endif()
+
+  if (NOT args_NO_INSTALL)
```

Comment:
Nit: Better call it `args_INSTALL` to avoid double-negation?

Reply:
It's not feasible. Default is to install and options for [`cmake_parse_arguments`](https://cmake.org/cmake/help/latest/command/cmake_parse_arguments.html) are True when provided and False otherwise. So it has to be `NO_INSTALL`. This shouldn't be a problem for users though, double negation here is an implementation detail.

Reply:
Ok, thanks for the details.